iquip/apps/explorer.py

```diff
@@ -1,32 +1,94 @@
 """App module for showing the experiment list and opening an experiment."""
 
-from typing import Optional, Tuple
+import os
+from typing import Optional, Tuple, Union
 
-from PyQt5.QtCore import QObject
-from PyQt5.QtWidgets import QLabel, QVBoxLayout, QWidget
+from PyQt5.QtCore import QObject, pyqtSlot
+from PyQt5.QtWidgets import (
+    QPushButton, QTreeWidget, QTreeWidgetItem, QVBoxLayout, QWidget
+)
 
 import qiwis
 
+from iquip import cmdtools
+
 class ExplorerFrame(QWidget):  # pylint: disable=too-few-public-methods
     """Frame for showing the experiment list and opening an experiment."""
 
     def __init__(self, parent: Optional[QWidget] = None):
         """Extended."""
         super().__init__(parent=parent)
         # widgets
-        self.testLabel = QLabel("explorer", self)
+        self.expStructure = QTreeWidget(self)
+        self.expStructure.header().setVisible(False)
+        self.reloadButton = QPushButton("Reload", self)
+        self.openButton = QPushButton("Open", self)
         # layout
         layout = QVBoxLayout(self)
-        layout.addWidget(self.testLabel)
+        layout.addWidget(self.reloadButton)
+        layout.addWidget(self.expStructure)
+        layout.addWidget(self.openButton)
 
 
 class ExplorerApp(qiwis.BaseApp):  # pylint: disable=too-few-public-methods
     """App for showing the experiment list and opening an experiment."""
 
-    def __init__(self, name: str, parent: Optional[QObject] = None):
+    def __init__(self, name: str, masterPath: str = ".", parent: Optional[QObject] = None):
         """Extended."""
         super().__init__(name, parent=parent)
+        self.repositoryPath = os.path.join(masterPath, "repository")
         self.explorerFrame = ExplorerFrame()
+        self.loadExpStructure()
+        # connect signals to slots
+        self.explorerFrame.reloadButton.clicked.connect(self.loadExpStructure)
+        self.explorerFrame.openButton.clicked.connect(self.openExp)
+
+    @pyqtSlot()
+    def loadExpStructure(self):
+        """Loads the experiment file structure in self.explorerFrame.expStructure.
+
+        It assumes that all experiment files are in self.repositoryPath.
+        """
+        self.explorerFrame.expStructure.clear()
+        self._addExpFile(self.repositoryPath, self.explorerFrame.expStructure)
+
+    def _addExpFile(self, path: str, parent: Union[QTreeWidget, QTreeWidgetItem]):
+        """Searches the sub elements and add them into self.explorerFrame.expStructure.
+
+        This uses the DFS algorithm.
+        1. Fetch the command result, which is a list of sub elements.
+        2. If a directory, call _addExpFile() recursively.
+        3. Otherwise, add it to self.explorerFrame.expStructure.
+        """
+        expList = cmdtools.run_command(f"artiq_client ls {path}").stdout
```

Comment:
Oh, I missed that part!
How about catching an exception in `cmdtools.run_command()` and in such case, returning `None`.
It looks more convenient for using this function.

Reply:
It will be convenient, but it loses the error information.
Maybe we can add a parameter to `run_command()` to return or raise the exception or just ignore it.
However, in that case, it seems using `subprocess` itself is better..
If you think using `cmdtools` is still better, you can just ignore the exception for now. Later we have to update the API when we need exception information.

Reply:
Umm.. Let's ignore it for now and improve it later!
I don't think we need to consider it much yet.